fuse/dl/cross_validation/pipeline.py

```diff
@@ -0,0 +1,303 @@
+        for training and 20% for validation, set `num_folds=5` and `num_folds_used=1`.
+        For running a 5-fold cross-validation scenario, set `num_folds=5` and `num_folds_used=5`.
+    :param num_gpus_total: Number of GPUs to use in total for executing the pipeline.
+    :param num_gpus_per_split: Number of GPUs to use for a single model training/inference.
+    :param num_repetitions: Number of repetitions of the procedure with different random
+        seeds. Note that this does not change the random decision on cross validation
+        fold sample ids.
+    :param dataset_func: Callable to a custom function that implements a dataset
+        creation. Its input is a path to cache directory and it should return a
+        train and test dataset.
+    :param train_func: Callable to a custom function that executes model training.
+    :param infer_func: Callable to a custom inference function.
+    :param eval_func: Callable to a custom evaluation function.
+    :param dataset_params: Dictionary that can contain any number of additional
+        custom parameters for dataset_func.
+    :param train_params: Dictionary that can contain any number of additional
+        custom parameters for train_func.
+    :param infer_params: Dictionary that can contain any number of additional
+        custom parameters for infer_func.
+    :param eval_params: Dictionary that can contain any number of additional
+        custom parameters for eval_func.
+    :param deterministic_mode: Whether to use PyTotch deterministic mode.
+    :param sample_ids_per_fold: May contain a sequence of array pairs denoting sample ids
+        for pre-defined train/validation splits. If this parameter is set to `None`,
+        the splits are decided at random.
+    """
+
+    if deterministic_mode:
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"  # required for pytorch deterministic mode
+
+    available_gpu_ids = FuseUtilsGPU.get_available_gpu_ids()
+    if num_gpus_total < len(available_gpu_ids):
+        available_gpu_ids = available_gpu_ids[0:num_gpus_total]
+    # group gpus into chunks of size num_gpus_per_split
+    gpu_resources = [
+        available_gpu_ids[i : i + num_gpus_per_split] for i in range(0, len(available_gpu_ids), num_gpus_per_split)
+    ]
+
+    # create a queue of gpu chunks (resources)
+    q_resources = Queue()
+    for r in gpu_resources:
+        q_resources.put(r)
+
+    if sample_ids_per_fold is None:
+        dataset, test_dataset = dataset_func(paths=paths, params=dataset_params)
+        # the split decision should be the same regardless of repetition index
+        kfold = KFold(n_splits=num_folds, shuffle=True, random_state=1234)
+        sample_ids_per_fold = [item for item in kfold.split(dataset)]
+    else:
+        assert num_folds == len(sample_ids_per_fold)
+
+    for rep_index in range(num_repetitions):
+        # run training, inference and evaluation on all cross validation folds in parallel
```

Comment:
I think that one step towards multiple computing nodes would be to store all the arguments for a single run in the filesystem and execute it with a new python program.

Reply:
let's discuss this when we get to extending it to multiple nodes